### PR TITLE
Feature/network

### PR DESCRIPTION
**Pull request by ayzhnau2013**
## Overview
add network module
## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## In scope
retrofit module
retrofit demo
## Check list
1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally prior to submission?
3. [ ] Have you update the wiki of your change?

## Reference
- Card: Should add the card url on here.
- Related PR: The related pr of current one.

### CODE REVIEW

File: app/src/main/java/com/thoughtworks/android/ark/ui/home/HomeFragment.kt [hunk 1]
```diff
@@ -33,6 +33,7 @@ class HomeFragment : Fragment() {
         homeViewModel.text.observe(viewLifecycleOwner) {
             textView.text = it
         }
+        homeViewModel
```
[the remark this answers is not in the document]
Comment (ayzhnau2013) [reply app/src/main/java/com/thoughtworks/android/ark/ui/home/HomeFragment.kt:36]: ok

File: app/build.gradle [hunk 2]
```diff
@@ -73,6 +74,8 @@ dependencies {
     implementation "com.google.dagger:hilt-android:$hilt_version"
     kapt "com.google.dagger:hilt-android-compiler:$hilt_version"
 
+    implementation 'com.squareup.retrofit2:retrofit:2.6.0'
```
Comment (Tony---Zhang) [inline app/build.gradle:77]: 2.6.0  version need to set to `ext {}` in root `build.gradle`

File: app/src/main/java/com/thoughtworks/android/ark/ui/network/UserRepository.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.thoughtworks.network.entity.RetrofitResponse
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import javax.inject.Inject
+
+class UserRepository @Inject constructor(private val remoteDataApi: RemoteDataApi) {
+    suspend fun getData() : Flow<RetrofitResponse> {
+        return flow {
+            emit(RetrofitResponse.Loading)
```
Comment (Tony---Zhang) [inline app/src/main/java/com/thoughtworks/android/ark/ui/network/UserRepository.kt:11]: Loading state is better to move to viewmodel

File: core/src/main/java/com/thoughtworks/android/core/network/entity/NetworkConfig.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.thoughtworks.android.core.network.entity
+
+data class NetworkConfig(
```
Comment (yunlongwen) [inline core/src/main/java/com/thoughtworks/android/core/network/entity/NetworkConfig.kt:3]: `NetworkConfig` is never been used.
Should it be ready to be used in `RetrofitClient#createService`?

Comment (ayzhnau2013) [reply core/src/main/java/com/thoughtworks/android/core/network/entity/NetworkConfig.kt:3]: This class will be used in ApiService init function.But it need to be called by integrator.I think this class is used in someone so I don't remove it.This class can also be deleted, and the configuration items are left to the user to configure.I will remove this class.Thanks

File: app/src/main/java/com/thoughtworks/android/ark/ui/home/HomeViewModel.kt [hunk 5]
```diff
@@ -1,19 +1,43 @@
...
 ) : ViewModel() {
 
     private val _text = liveData {
         emit(repository.loadData())
     }
 
     val text: LiveData<String> = _text
+
+    private val _state = MutableLiveData("")
+    val state:LiveData<String>
```
Comment (Tony---Zhang) [inline app/src/main/java/com/thoughtworks/android/ark/ui/home/HomeViewModel.kt:23]: 这里少了一个空格吧？val state: LiveData<String>

File: app/src/main/java/com/thoughtworks/android/ark/ui/network/DataEntity.kt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com.thoughtworks.android.ark.ui.network
+
+data class DataEntity(
```
Comment (Tony---Zhang) [inline app/src/main/java/com/thoughtworks/android/ark/ui/network/DataEntity.kt:3]: change class name to `GithubUserEntity`?

File: core/src/main/java/com/thoughtworks/android/core/network/callback/RetrofitCallback.kt [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import java.net.ConnectException
+import java.net.SocketTimeoutException
+
+abstract class RetrofitCallback<T:Any> : Callback<BaseCallModel<T>>  {
+
+    override fun onResponse(call: Call<BaseCallModel<T>>, response: Response<BaseCallModel<T>>) {
+        when(response.raw().code()) {
+            200 -> {
+                when(response.body()?.errorCode) {
+                    0 -> onSuccess(response.body()!!.data)
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/callback/RetrofitCallback.kt:16]: response.body()?.let { onSuccess(it.data) }

File: core/src/main/java/com/thoughtworks/android/core/network/callback/RetrofitCallback.kt [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+                    else -> response.body()?.errorMsg?.let { onFailed(it) }
+                }
+            }
+            else -> onFailure(call, RuntimeException("response error"))
+        }
+    }
+
+    override fun onFailure(call: Call<BaseCallModel<T>>, t: Throwable) {
+        val errorMsg = when (t) {
+            is SocketTimeoutException -> "sock exception"
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/callback/RetrofitCallback.kt:26]: HttpTimeOutException? -> "http timeout exception"

File: core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import android.content.Context
+import com.thoughtworks.android.core.network.client.http.DefaultHttpClient
+import com.thoughtworks.android.core.network.client.retrofit.BaseRetrofit
+import com.thoughtworks.android.core.network.client.retrofit.DefaultRetrofit
+import retrofit2.Retrofit
+
+class RetrofitClient(private val context: Context) {
+    private lateinit var httpClient: DefaultHttpClient
+    private lateinit var retrofitClient: BaseRetrofit
+    private val retrofitMap:MutableMap<String,Retrofit> = hashMapOf()
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt:12]: private val retrofitMap: MutableMap<String,Retrofit> = hashMapOf()

File: core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    private fun initRetrofit(baseUrl:String = HOST_URL) {
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt:14]: private fun initRetrofit(baseUrl: String = HOST_URL) {

File: core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        retrofitMap.clear()
+    }
+
+    fun <T:BaseRetrofit> setupCustomRetrofit(retrofitClient: T) {
+        this.retrofitClient = retrofitClient
+        retrofitMap.clear()
+    }
+
+    companion object {
+        const val HOST_URL = "http://www.baidu.com"
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt:39]: Delete this host url

File: core/src/main/java/com/thoughtworks/android/core/network/client/http/DefaultHttpClient.kt [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+import okhttp3.OkHttpClient
+import okhttp3.logging.HttpLoggingInterceptor
+import java.util.concurrent.TimeUnit
+
+open class DefaultHttpClient(private val context: Context) {
+    var okHttpClient:OkHttpClient
+
+    init {
+        okHttpClient = OkHttpClient.Builder().apply {
+            connectTimeout(TIME_OUT * 3, TimeUnit.SECONDS)
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/client/http/DefaultHttpClient.kt:15]: Why here * 3? in BaseHttpClient dont have this logic

File: core/src/main/java/com/thoughtworks/android/core/network/client/http/BaseHttpClient.kt [hunk 13]
```diff
@@ -0,0 +1,23 @@
+package com.thoughtworks.android.core.network.client.http
+
+import okhttp3.OkHttpClient
+import java.util.concurrent.TimeUnit
+
+abstract class BaseHttpClient {
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/client/http/BaseHttpClient.kt:6]: Delete this `BaseHttpClient`

File: core/src/main/java/com/thoughtworks/android/core/network/interceptor/NetConnectInterceptor.kt [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+import okhttp3.Response
+import java.io.IOException
+
+class NetConnectInterceptor(private val context: Context) : Interceptor {
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val builder = chain.request().newBuilder()
+        if(hasNetworkConnect(context)) {
+            return chain.proceed(builder.build())
+        } else {
+            Log.e("tag","network is not connect")
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/interceptor/NetConnectInterceptor.kt:16]: Delete this `Log.e`

File: core/src/main/java/com/thoughtworks/android/core/network/util/NetworkUtil.kt [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+private suspend fun <T> getResponse(apiCall: suspend () -> Response<T>) : RetrofitResponse<T> {
+    return try {
+        val response = apiCall()
+        if(response.isSuccessful) {
+            RetrofitResponse.Success(response.body()!!)
+        } else {
+            RetrofitResponse.Error("result is error.")
+        }
+    } catch (e:SocketTimeoutException) {
+        RetrofitResponse.Error("socket time out exception")
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/util/NetworkUtil.kt:31]: "socket time out exception" from Error.message?

File: core/src/main/java/com/thoughtworks/android/core/network/util/NetworkUtil.kt [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    }
+}
+
+fun hasNetworkConnect(context: Context): Boolean {
+    var isConnect = false
+    val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+    connectivityManager.run {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            getNetworkCapabilities(activeNetwork)?.run {
+                isConnect = when {
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/util/NetworkUtil.kt:47]: isConnect = hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
                            hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) ||
                            hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)

File: core/src/main/java/com/thoughtworks/android/core/network/util/NetworkUtil.kt [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+                    hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
+                            hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) ||
+                            hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)    -> true
+                    else -> false
+                }
+            }
+        } else {
+            activeNetworkInfo?.run {
+                isConnect = when(type) {
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/util/NetworkUtil.kt:56]: isConnect = type == ConnectivityManager.TYPE_WIFI || type == ConnectivityManager.TYPE_MOBILE

File: app/src/main/java/com/thoughtworks/android/ark/ui/home/feeds/FeedScreen.kt [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+) {
+    Column(modifier = Modifier.fillMaxSize()) {
+        Spacer(modifier = Modifier.height(10.dp))
+        Button(
+            onClick = {
+               dispatcherAction(FeedUiAction.FriendListAction)
+            },
+            modifier = Modifier
+                .fillMaxWidth()
+                .height(50.dp)
```
Comment (Tony---Zhang) [inline app/src/main/java/com/thoughtworks/android/ark/ui/home/feeds/FeedScreen.kt:43]: 50 -> 48dp
5 -> 4dp

File: app/src/main/res/layout/fragment_home.xml [hunk 19]
```diff
@@ -15,8 +15,14 @@
...
-        app:layout_constraintEnd_toEndOf="parent"
         app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toTopOf="parent" />
+        app:layout_constraintTop_toTopOf="parent"/>
+
+    <androidx.compose.ui.platform.ComposeView
+        android:id="@+id/home_compose"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        tools:ignore="MissingConstraints"
```
Comment (Tony---Zhang) [inline app/src/main/res/layout/fragment_home.xml:25]: 这个布局有问题吧？

File: core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+import android.content.Context
+import com.thoughtworks.android.core.network.client.http.DefaultHttpClient
+import com.thoughtworks.android.core.network.client.retrofit.BaseRetrofit
+import com.thoughtworks.android.core.network.client.retrofit.DefaultRetrofit
+import retrofit2.Retrofit
+
+class RetrofitClient(private val context: Context) {
+    private lateinit var httpClient: DefaultHttpClient
+    private lateinit var baseRetrofit: BaseRetrofit
+    private val retrofitMap:MutableMap<String, Retrofit> = hashMapOf()
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt:12]: private val retrofitMap:MutableMap<String, Retrofit> = hashMapOf()
private val retrofitMap: MutableMap<String, Retrofit> = hashMapOf()

File: core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+    private fun initRetrofit(baseUrl: String) {
+        httpClient = DefaultHttpClient(context)
+        baseRetrofit = DefaultRetrofit()
+        val retrofit = baseRetrofit.createRetrofit(baseUrl, httpClient.okHttpClient)
+        retrofitMap[baseUrl] = retrofit
+    }
+
+    fun <T> createService(clazz:Class<T>, baseUrl: String):T {
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt:21]: 这个文件format 一下

File: core/src/main/java/com/thoughtworks/android/core/network/entity/NetConnectException.kt [hunk 22]
```diff
@@ -0,0 +1,5 @@
+package com.thoughtworks.android.core.network.entity
+
+import java.io.IOException
+
+class NetConnectException(msg: String?) : IOException(msg)
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/entity/NetConnectException.kt:5]: NetworkConnectException

File: core/src/main/java/com/thoughtworks/android/core/network/interceptor/BaseInterceptor.kt [hunk 23]
```diff
@@ -0,0 +1,16 @@
+package com.thoughtworks.android.core.network.interceptor
+
+import okhttp3.Interceptor
+import okhttp3.Response
+
+open class BaseInterceptor(private val headers:Map<String,String>?) : Interceptor {
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/interceptor/BaseInterceptor.kt:6]: open class BaseInterceptor(private val headers: Map<String,String>?) : Interceptor {

File: core/src/main/java/com/thoughtworks/android/core/network/interceptor/NetConnectInterceptor.kt [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+import android.content.Context
+import com.thoughtworks.android.core.network.entity.NetConnectException
+import com.thoughtworks.android.core.network.util.hasNetworkConnect
+import okhttp3.Interceptor
+import okhttp3.Response
+
+class NetConnectInterceptor(private val context: Context) : Interceptor {
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val builder = chain.request().newBuilder()
+        if(hasNetworkConnect(context)) {
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/interceptor/NetConnectInterceptor.kt:12]: if (hasNetworkConnect(context)) {

File: core/src/main/java/com/thoughtworks/android/core/network/util/NetworkUtil.kt [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    }
+}
+
+private suspend fun <T> getResponse(apiCall: suspend () -> Response<T>): RetrofitResponse<T> {
+    return try {
+        val response = apiCall()
+        if (response.isSuccessful) {
+            RetrofitResponse.Success(response.body()!!)
+        } else {
+            RetrofitResponse.Error(0,"result is error.")
```
Comment (Tony---Zhang) [inline core/src/main/java/com/thoughtworks/android/core/network/util/NetworkUtil.kt:35]: RetrofitResponse.Error(0, "result is error.")

**Review by Tony---Zhang — APPROVED** [review]

File: core/src/main/java/com/thoughtworks/android/core/network/client/http/DefaultHttpClient.kt [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+
+import android.content.Context
+import com.thoughtworks.android.core.BuildConfig
+import com.thoughtworks.android.core.network.interceptor.NetConnectInterceptor
+import okhttp3.OkHttpClient
+import okhttp3.logging.HttpLoggingInterceptor
+import java.util.concurrent.TimeUnit
+
+open class DefaultHttpClient(private val context: Context) {
+    var okHttpClient: OkHttpClient
```
Comment (jie-meng) [inline core/src/main/java/com/thoughtworks/android/core/network/client/http/DefaultHttpClient.kt:11]: If okHttpClient should not access by other class, it should be private

File: core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+import android.content.Context
+import com.thoughtworks.android.core.network.client.http.DefaultHttpClient
+import com.thoughtworks.android.core.network.client.retrofit.BaseRetrofit
+import com.thoughtworks.android.core.network.client.retrofit.DefaultRetrofit
+import retrofit2.Retrofit
+
+class RetrofitClient(private val context: Context) {
+    private lateinit var httpClient: DefaultHttpClient
+    private lateinit var baseRetrofit: BaseRetrofit
+    private val retrofitMap: MutableMap<String, Retrofit> = hashMapOf()
```
Comment (jie-meng) [inline core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt:12]: Should here use mutableMapOf() ?

File: core/src/main/java/com/thoughtworks/android/core/network/client/http/DefaultHttpClient.kt (hunk 26 above)
Comment (ayzhnau2013) [reply core/src/main/java/com/thoughtworks/android/core/network/client/http/DefaultHttpClient.kt:11]: this attribute was used in retrofit client.So this attribute is not declared private .

File: core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt (hunk 27 above)
Comment (ayzhnau2013) [reply core/src/main/java/com/thoughtworks/android/core/network/client/RetrofitClient.kt:12]: mutable map is a interface and hash map implement this interface.I think there is no problem in this way.It is no problem to use hashmap directly.

**Comment by jie-meng** [comment]
👍

**Review by jie-meng — APPROVED** [review]

**Pull request merged**